mon,osd: improve mon scalability by vastly simplyfing the osd <-> mon communications

liewegas opened the pull request:
* mon: waste less cpu iterating over stale pgs
* osdmap: cache up/in values
* osd: exponential backoff of mon reports
* make subscribe fully stateful, no renewals
* mon: let peons send osdmap replies
* mon: redo pg creations
* osd: no dup alive, failure, or pg_temp messages
* osd: explicit preboot stage
* mon: use keepalive (not subscribe) for liveness

This made a mon cluster for 2000 osds pretty hard to break (call elections) in spite
of creating and rebalancing huge pools.

gregsfortytwo commented on src/mon/PGMonitor.cc:
```diff
@@ -1220,28 +1273,36 @@ bool PGMonitor::check_down_pgs()
-	stat = &q->second;
+  // if a large number of osds changed state, just iterate over the whole
+  // pg map.
+  if (need_check_down_pg_osds.size() > (unsigned)osdmap->get_max_osd() / 20)
```
Should make that 20 a config value — I guarantee we will build somebody a custom binary for it, otherwise. ;)
(Unless there's some math to pick out that 5% that I'm missing?)

Also, shouldn't we be looking at get_num_osds() instead of the max osd ID in use?

gregsfortytwo commented on src/osd/OSDMap.cc:
```diff
@@ -937,9 +937,18 @@ void OSDMap::set_max_osd(int m)
+      ++num_up_osd;
+    if ((osd_state[i] & CEPH_OSD_EXISTS) &&
+	get_weight(i) != CEPH_OSD_OUT)
+      ++num_in_osd;
```
Not terribly important (and probably optimized away) but why not nest these instead of double-testing the exists check?

gregsfortytwo commented on src/osd/OSD.cc:
```diff
@@ -4036,6 +4061,64 @@ void OSD::tick_without_osd_lock()
+      last_pg_stats_ack = ceph_clock_now(cct);  // reset clock
+      last_pg_stats_sent = utime_t();
+      stats_ack_timeout = MAX(g_conf->osd_mon_ack_timeout,
+			      stats_ack_timeout * 2.0);
```
Should be a configurable multiple.

gregsfortytwo commented on src/osd/OSD.cc:
```diff
@@ -4795,10 +4887,19 @@ void OSD::handle_pg_stats_ack(MPGStatsAck *ack)
+
+  // decay timeout slowly (analogous to TCP)
+  stats_ack_timeout = MAX(g_conf->osd_mon_ack_timeout,
+			  stats_ack_timeout * .9);
```
Should be a configurable multiple.

gregsfortytwo commented on src/common/config_opts.h:
```diff
@@ -648,8 +649,9 @@ OPTION(osd_pg_max_concurrent_snap_trims, OPT_U64, 2)
-OPTION(osd_mon_report_interval_max, OPT_INT, 120)
+OPTION(osd_mon_report_interval_max, OPT_INT, 600)
 OPTION(osd_mon_report_interval_min, OPT_INT, 5)  // pg stats, failures, up_thru, boot.
+OPTION(osd_mon_report_max_in_flight, OPT_INT, 2)  // max updates in flight
```
This value looks to only be applied in between the min and max intervals; that should be documented.

gregsfortytwo commented on src/osd/OSD.cc:
```diff
@@ -6539,10 +6643,13 @@ void OSD::handle_osd_map(MOSDMap *m)
-    start_boot();  // retry
+  else if (is_preboot()) {
+    if (m->get_source().is_mon())
+      _preboot(m->oldest_map, m->newest_map);
```
This doesn't work. The message's oldest_map is set based on the maps the OSD already has, not necessarily the oldest map on the monitor. "send_latest(op, m->sb.current_epoch+1);"

liewegas replied:
https://github.com/ceph/ceph/blob/master/src/mon/OSDMonitor.cc#L2432

it's the newest map on the sender.. not related to the maps being sent.

gregsfortytwo replied:
oldest_map is the one I'm concerned about, not newest. That's filled in to the newer of our oldest map, or the monitor's oldest map, but we treat it as the monitor's oldest in _preboot. (Granted it looks to be mainly debug output in there, but it will be confusing. I had it in my head we were doing some osdmap trims based on these values but now I'm not seeing it, so not sure if there's a more important use or not.)

liewegas replied:
Oh, yep.. it looks like that site is just wrong (everyone else who creates an MOSDMap fills it in with the oldest map on the sender).

gregsfortytwo commented on src/osd/OSD.cc:
```diff
@@ -4332,20 +4415,6 @@ void OSD::RemoveWQ::_process(
-  service.send_pg_temp();
-  send_failures();
-  send_pg_stats(now);
-}
```
Why do you want to inline do_mon_report?

gregsfortytwo commented on src/osd/OSD.cc:
```diff
@@ -4332,20 +4415,6 @@ void OSD::RemoveWQ::_process(
-  last_mon_report = now;
-
-  // do any pending reports
-  send_alive();
```
You seem to have ditched this call to send_alive but I can't find anything that replaced it in the same part of the git timeline.

liewegas replied:
It's already called in queue_want_up_thru()

gregsfortytwo commented on src/osd/OSD.cc:
```diff
@@ -4703,24 +4779,38 @@ void OSD::got_full_map(epoch_t e)
+    if (!failure_pending.count(osd)) {
+      int failed_for = (int)(double)(now - failure_queue.begin()->second);
+      monc->send_mon_message(new MOSDFailure(monc->get_fsid(), i, failed_for,
+					     osdmap->get_epoch()));
```
I don't think this filtering is correct. The monitors distinguish between the number of failure reports and the number of failure reporters (the first being larger than the second) and that tends to matter for small clusters.

gregsfortytwo commented:
Sorry for the duplicate comments, but I don't want to lose the record if commits get rebased.

gregsfortytwo commented on src/mon/MonClient.cc:
```diff
@@ -709,28 +715,30 @@ void MonClient::tick()
     cur_con->send_keepalive();
 
     if (state == MC_STATE_HAVE_SESSION) {
-      send_log();
```
Why don't you want to send the log if reconnecting? The information in it might still be useful.
(Commit "mon/MonClient: don't send log if we're reconnecting")

gregsfortytwo commented:
Commit "msg/simple/Pipe: show keepalives at level 2": did you mean to leave that around? I guess I'm not opposed.

gregsfortytwo commented on src/mon/Monitor.cc:
```diff
@@ -3271,6 +3271,11 @@ void Monitor::handle_route(MonOpRequestRef op)
+	dout(10) << " sending osdmaps from " << m->send_osdmap_first << dendl;
+	osdmon()->send_incremental(m->send_osdmap_first, session,
+				   true, MonOpRequestRef());
+      }
```
I don't think it's required, but the name "send_osdmap_first" definitely makes me think we're sending the OSDMaps back before the normal response. And given the code structure I think that's the more likely pattern if we add anything that does require a separate response besides the maps.

gregsfortytwo commented:
Commit "mon/PGMonitor: only map and send pg creates post paxos update" means we have to wait through another round trip to start creating PGs (instead of the one it takes to update the OSDMap). Might be fine but I'd be surprised if the map iteration was expensive enough to be worth it.

gregsfortytwo commented on src/mon/MonClient.h:
```diff
@@ -209,31 +209,42 @@ class MonClient : public Dispatcher {
-      if (sub_have[what].flags & CEPH_SUBSCRIBE_ONETIME)
-	sub_have.erase(what);
+    if (sub_new.count(what)) {
+      if (sub_new[what].flags & CEPH_SUBSCRIBE_ONETIME)
```
Is it okay that we're dropping all onetime subscriptions, regardless of the requested epoch values?

liewegas replied:
fixed

gregsfortytwo commented on src/osd/OSD.cc:
```diff
@@ -1965,7 +2000,15 @@ int OSD::init()
+  monc->sub_want("osd_pg_creates", last_pg_create_epoch, 0);
+
+  // we don't need to ask for an osdmap here; objecter will
+  //monc->sub_want("osdmap", osdmap->get_epoch(), CEPH_SUBSCRIBE_ONETIME);
```
Hmm, this works as long as we keep running the objecter unconditionally. I guess that works, but can't we just leave it asking and let the MonClient merge requests?

liewegas replied:
I tried that first but had trouble getting it to send a single subscribe instead of multiples.  But it might be possible later.. hence leaving the ugly comment in the code.

gregsfortytwo replied:
Oh, because these are ONETIME subscribes and get cleaned out. Yeah, okay.

gregsfortytwo commented on src/mon/MonClient.cc:
```diff
@@ -764,14 +772,19 @@ void MonClient::_renew_subs()
 
     MMonSubscribe *m = new MMonSubscribe;
-    m->what = sub_have;
+    m->what = sub_new;
```
This doesn't work, we need to combine sub_new and sub_sent whenever we _renew_subs(), or else we aren't including everything. We have to do that whenever we connect to a new monitor, and this is the path for that...
(See the osd_pg_creates subscription change in OSD::init for just one example of how this will otherwise break things.)

liewegas replied:
_reopen_session merges sub_sent back into _sub_new.

gregsfortytwo replied:
Mmm, so it does. Are we okay not resending existing subscriptions if we're talking to an older monitor?

gregsfortytwo commented on src/mon/PGMonitor.cc:
```diff
@@ -1132,47 +1139,58 @@ void PGMonitor::map_pg_creates()
+      dout(20) << __func__ << "  " << pgid << " "
+	       << " primary: " << s->acting_primary << " -> " << acting_primary
+	       << " acting: " << s->acting << " -> " << acting
+	       << dendl;
```
We should print out the up[_primary] as well, since it can lead us into this branch?
Also the spacing doesn't seem to match on these lines.

gregsfortytwo commented:
Okay, finally got time to finish up the last few commits.
@tchaikov, @jecluis, you probably both want to check over this as well?

tchaikov commented:
@gregsfortytwo thanks. yeah, will do!

liewegas commented:
- had to rebase this, sorry!
- fixed pg creations.  after going around in circles replicating a bunch of peering-like code i realized peering did all this and instead got to rip out a ton of old cruft, yay!
- fixed the reporters vs reported issue.

Need to run this through some thrashing now to make sure the pg creation logic is really as simple as I think it is!

tchaikov commented:
finished review. in general, lgtm.

liewegas commented:
merged manually
